### DISCUSSION TITLE
Show user location independently of `GeolocateControl`

### DISCUSSION THREAD

**Discussion by JulianBissekkou** [maplibre/maplibre-gl-js#2960] [Ideas]
## User Story

As an developer I would like to integrate the mapbox location easier into the location business logic of my product by updating the location indicator independently from the `GeolocationControl`

## Rationale

Its a common use case for apps to load the location of the user independently from the map to do other business logic with it. If you decide to give the user more context you can show a map but this is not required.

For example we might have a location aware search that filters nearby shops based on the distance to the user. The user can search and find places, but we only want to show the map with the users location when the user presses the option "show on map".
It woud be nice to give the map the location that we fetched previously and show it on the map without interacting with the `GeolocationControl`.
Right now the UX is very poor because we have to open the map and wait for the `GeolocationControl` to fetch the location again. 
Afaik there is not workaround for this. All location updates need to be performed by `GeolocationControl`.

**Comment by HarelM** [comment]
I'm not sure I understand the requirement here.
The location control is a simple out of the box solution to show the user location.
If you would like to build something else, more complex, you should create a custom control or just add a location logic to you app...
You can use the browser navigator API to accommodate for your needs I guess...

**Comment by JulianBissekkou** [comment]
In order to do so I would like to reuse the location indicator that shows the user location on the map, otherwise we have to build that again by ourselves.

**Comment by HarelM** [comment]
I would recommend looking at the code here and understanding how it works and tweak it as you need.
https://github.com/maplibre/maplibre-gl-js/blob/main/src/ui/control/geolocate_control.ts

Since I can't understand the requirements here in order to actually define a behavior that is missing...

**Comment by stefanschaller** [comment]
@HarelM I'm working together with @JulianBissekkou. 

We already have a location stream that changes whenever the position of the user has changed. 
So what we want is basically to draw the user location indication by our own. 
We want to do that because of the following reasons: 

- Reuse the existing current location that we already use for other features in our app. 
- Avoid the behavior that the user need to manually click the "current location button" of the map

**Comment by HarelM** [comment]
As stated before, I don't understand the requirements to the location control.
What feature is missing? What would you like to add or change? What problem should it solve?

**Comment by mrtnrs** [comment]
I guess the thing is that it would be nice to be able to handle the location fetching part ourselves; so we could plug our own location data into GeolocationControl to use its functionality/styling. As right now we need to either build/copy the entire functionality/styling ourselves; or use parts and copy for example the styling into our own setup.

**Comment by HarelM** [comment]
Yea, I don't see this as a general request that other can actually use.
Styling something in maplibre is relatively easy.

**Comment by ferdicus** [comment]
if i interpreted OP correctly, i think what they meant was: being able to show the user location (and the marker on the map), independent of having to use or show `GeolocateControl`

as it is right now:
* show `GeolocateControl` on load
* user has to click control button 🎯 
* now the users location is shown/ tracked

the feature request seems to be:
* programmatically enable user tracking (no control being shown/ no `GeolocateControl` used)
* users location is automatically shown on map load - no button click required

i myself was confused, that this is not described in the docs and when searching user location only this shows up: https://maplibre.org/maplibre-gl-js/docs/examples/locate-user/

which links tracking a users location to the ui control element 🤔

**Reply by maker-chrobok** [reply]
Use trigger(false) this programmatically request and move the map to the user's location.

`// Initialize the geolocate control.
let geolocate = new GeolocateControl({
 positionOptions: {
   enableHighAccuracy: true
 },
 trackUserLocation: true
});
// Add the control to the map.
map.addControl(geolocate);
map.on('load', () => {
  geolocate.trigger(false);
});`

https://maplibre.org/maplibre-gl-js/docs/API/classes/GeolocateControl/#trigger

**Comment by HarelM** [comment]
I'm conversing this into a discussion as I don't see a real value in keeping this issue open.
I have yet to receive anything concrete that worth implementing.
If there's something specific I would suggest to open a pull request.
In my app for example, I use a custom geolocation control I wrote as I needed custom functionally:
https://github.com/IsraelHikingMap/Site/blob/451c515c06d0fe96b8ad4f19791eae9496235f1f/IsraelHiking.Web/src/application/components/location.component.ts
